firebase-firestore/src/main/java/com/google/firebase/firestore/Query.java

```diff
@@ -553,44 +603,71 @@ private void validateOrderByFieldMatchesInequality(
-        }
-        com.google.firebase.firestore.model.FieldPath firstOrderByField =
-            query.getFirstOrderByField();
-        if (firstOrderByField != null) {
-          validateOrderByFieldMatchesInequality(firstOrderByField, newInequality);
-        }
+      if (existingInequality != null && !existingInequality.equals(newInequality)) {
+        throw new IllegalArgumentException(
+            String.format(
+                "All where filters with an inequality (notEqualTo, notIn, lessThan, "
+                    + "lessThanOrEqualTo, greaterThan, or greaterThanOrEqualTo) must be on the "
+                    + "same field. But you have filters on '%s' and '%s'",
+                existingInequality.canonicalString(), newInequality.canonicalString()));
       }
-      Operator conflictingOp = query.findFilterOperator(conflictingOps(filterOp));
-      if (conflictingOp != null) {
-        // We special case when it's a duplicate op to give a slightly clearer error message.
-        if (conflictingOp == filterOp) {
-          throw new IllegalArgumentException(
-              "Invalid Query. You cannot use more than one '" + filterOp.toString() + "' filter.");
-        } else {
-          throw new IllegalArgumentException(
-              "Invalid Query. You cannot use '"
-                  + filterOp.toString()
-                  + "' filters with '"
-                  + conflictingOp.toString()
-                  + "' filters.");
+      com.google.firebase.firestore.model.FieldPath firstOrderByField =
+          query.getFirstOrderByField();
+      if (firstOrderByField != null) {
+        validateOrderByFieldMatchesInequality(firstOrderByField, newInequality);
+      }
+    }
+    Operator conflictingOp = findFilterOperator(query.getFilters(), conflictingOps(filterOp));
+    if (conflictingOp != null) {
+      // We special case when it's a duplicate op to give a slightly clearer error message.
+      if (conflictingOp == filterOp) {
+        throw new IllegalArgumentException(
+            "Invalid Query. You cannot use more than one '" + filterOp.toString() + "' filter.");
+      } else {
+        throw new IllegalArgumentException(
+            "Invalid Query. You cannot use '"
+                + filterOp.toString()
+                + "' filters with '"
+                + conflictingOp.toString()
+                + "' filters.");
+      }
+    }
+  }
+
+  /** Checks that adding the given filter to the current query is valid */
+  private void validateAddingNewFilter(com.google.firebase.firestore.core.Filter filter) {
+    com.google.firebase.firestore.core.Query testQuery = query;
+    for (FieldFilter subfilter : filter.getAllFieldFilters()) {
+      validateAddingNewFieldFilter(testQuery, subfilter);
+      testQuery = query.filter(subfilter);
+    }
+  }
+
+  /**
+   * Checks if any of the provided filter operators are included in the given list of filters and
+   * returns the first one that is, or null if none are.
+   */
```

Comment:
This function was moved from firestore/core/Query.java to firestore/Query.java as it was the only consumer of it.